### PR TITLE
Add telemetry event in shared matrix for size of matrix

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Add telemetry event in shared matrix for size of matrix.

### CODE REVIEW

File: packages/dds/matrix/src/matrix.ts [hunk 1]
```diff
@@ -521,20 +523,34 @@ export class SharedMatrix<T = any>
 			SnapshotPath.cols,
 			this.cols.summarize(this.runtime, this.handle, serializer),
 		);
+		const cellsSnapshot = this.cells.snapshot();
+		const props: Record<string, number> = {
+			cellsSnapshotSize: cellsSnapshot.length,
+			rowCount: this.rowCount,
+			colCount: this.colCount,
+			setCellsCountFromLastSummary: this.setCellsCountFromLastSummary,
```
Comment (vladsud) [inline packages/dds/matrix/src/matrix.ts:531]: I'm not sure how much interesting is setCellsCountFromLastSummary. It's not just a function of user actions, but also a function of summary frequency. I have hard time imagining how we will use it in telemetry. Consider removing, unless you have good case for it.

File: packages/dds/matrix/src/matrix.ts [hunk 2]
```diff
@@ -521,20 +523,34 @@ export class SharedMatrix<T = any>
...
-			artifactsToSummarize.push(this.cellLastWriteTracker.snapshot());
+			const trackerMatrixSnapshot = this.cellLastWriteTracker.snapshot();
+			artifactsToSummarize.push(trackerMatrixSnapshot);
+			props.trackerMatrixSnapshotSize = trackerMatrixSnapshot.length;
 		}
 		builder.addBlob(
 			SnapshotPath.cells,
 			serializer.stringify(artifactsToSummarize, this.handle),
 		);
+		this.logger.sendTelemetryEvent({
```
Comment (vladsud) [inline packages/dds/matrix/src/matrix.ts:550]: I'm not sure this is applicable here but mention for completeness (and maybe you can see how to use it here).

You can avoid this event by leveraging ITelemetryContext object that is passed into summarizeCore() as a second argument. Summarizer process will automatically include it. That said, this only works as accumulation, i.e. that way we can record some stats about all shared matrixes that summarized on this path, but not individual DDSs.
For example, we can say that total number of cells across all such DDSs was X.

The benefit of using context object - there is only one single event per summary with all the properties collected from all summarized objects, i.e. efficiency in logging (and reduced frequency).

I do not see an easy way to leverage that mechanism here, unless we start using DDS Id to create unique records, which we should not do, due to IDs potentially having privacy concerns.

That said, this event you added might be too verbose. It all depends on a number of matrixes in a file and how often they get changed. I think that's fine for starters, but we will need to monitor it.

**Review by vladsud — APPROVED** [review]

File: packages/dds/matrix/src/matrix.ts (hunk 1 above)
Comment (jatgarg) [reply packages/dds/matrix/src/matrix.ts]: Yeah, I was dicey too whether to add this or not. Removing it.

File: packages/dds/matrix/src/matrix.ts (hunk 2 above)
Comment (jatgarg) [reply packages/dds/matrix/src/matrix.ts]: Tried using this approach. Please take a look again.

File: packages/dds/matrix/src/matrix.ts [hunk 3]
```diff
@@ -521,20 +524,31 @@ export class SharedMatrix<T = any>
...
+			const trackerMatrixSnapshot = this.cellLastWriteTracker.snapshot();
+			artifactsToSummarize.push(trackerMatrixSnapshot);
+			props.tmsSize = trackerMatrixSnapshot.length; // trackerMatrixSnapshotSize
 		}
 		builder.addBlob(
 			SnapshotPath.cells,
 			serializer.stringify(artifactsToSummarize, this.handle),
 		);
+		const stats =
+			telemetryContext?.get("fluid:SharedMatrix", "details")?.toString().concat("_") ?? "";
```
Comment (vladsud) [inline packages/dds/matrix/src/matrix.ts:550]: If I got it right, this joins multiple stringified JSON payloads with "_" as a separator, right?
I'm not sure this is possible to parse back - JSON payload could have same "_" symbols, it's impossible to differentiate between the two types of symbols.

File: packages/dds/matrix/src/matrix.ts [hunk 4]
```diff
@@ -521,20 +524,31 @@ export class SharedMatrix<T = any>
...
+			const trackerMatrixSnapshot = this.cellLastWriteTracker.snapshot();
+			artifactsToSummarize.push(trackerMatrixSnapshot);
+			props.tmsSize = trackerMatrixSnapshot.length; // trackerMatrixSnapshotSize
 		}
 		builder.addBlob(
 			SnapshotPath.cells,
 			serializer.stringify(artifactsToSummarize, this.handle),
 		);
+		if (telemetryContext?.push !== undefined) {
+			telemetryContext.push("fluid:SharedMatrix", "details", [JSON.stringify(props)]);
```
Comment (vladsud) [inline packages/dds/matrix/src/matrix.ts:550]: Curious, why it has to be stringified? Can it be same object? I.e. stringification to be done once, by context, on top of array of objects?

Comment (vladsud) [reply packages/dds/matrix/src/matrix.ts:550]: Down below, type is value: TelemetryEventPropertyType[]):, so that should be natural step, no?

File: packages/runtime/runtime-utils/src/summaryUtils.ts [hunk 5]
```diff
@@ -388,6 +389,28 @@ export class TelemetryContext implements ITelemetryContext {
...
+	 */
+	push(prefix: string, property: string, value: TelemetryEventPropertyType[]): void {
+		const prevValue = this.telemetry.get(`${prefix}${property}`);
+		if (typeof prevValue === "string") {
+			try {
+				const values = JSON.parse(prevValue) as TelemetryEventPropertyType[];
+				values.push(...value);
+				this.telemetry.set(`${prefix}${property}`, JSON.stringify(values));
+				return;
+			} catch (err) {}
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summaryUtils.ts:403]: I think we want to overwrite the value, or maybe convert to [<old value>, <new value>] state, not <old value" state (catch flow, i.e. we deal not with array)

File: packages/runtime/runtime-utils/src/summaryUtils.ts [hunk 6]
```diff
@@ -388,6 +389,28 @@ export class TelemetryContext implements ITelemetryContext {
...
 	}
 
+	/**
+	 * {@inheritDoc @fluidframework/runtime-definitions#ITelemetryContext.append}
+	 */
+	push(prefix: string, property: string, value: TelemetryEventPropertyType[]): void {
+		const prevValue = this.telemetry.get(`${prefix}${property}`);
+		if (typeof prevValue === "string") {
+			try {
+				const values = JSON.parse(prevValue) as TelemetryEventPropertyType[];
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summaryUtils.ts:399]: This dance with continuous stringification and parsing is very expensive. Can we simply story as as an array?
The telemetry pipeline knows how to deal with arrays (that support was added some time ago, though I do not see it reflected in type system, i.e. TelemetryBaseEventPropertyType has no arrays, but take a look at convertToBasePropertyTypeUntagged() - it deals with "object" type and there is comment about arrays), so I think we should be totally fine with having here something like Record<string, TelemetryBaseEventPropertyType >[]

Even better would be to fix telemetry type system to make it clear - arrays are accepted (though need to be very careful here - they are accepted by ITelemetryLogger, but not by ITelemetryBaseLogger. I thought we split types for these two interfaces (i.e. ITelemetryLogger should accepts more types and be able to convert them), but I'm not sure I see that.

I'd recommend tracing history and pinging who was added array types (sorry, I forgot who did it) - I think this work somewhat incomplete without this type fixup.

Also adding @markfields to review / validate my train of thought

**Review by vladsud — APPROVED** [review]
I believe we should not strigify arrays. Otherwise looks good!

File: packages/dds/matrix/src/matrix.ts (hunk 4 above)
Comment (markfields) [reply packages/dds/matrix/src/matrix.ts:550]: We took a step of supporting single-level arrays and objects. Object nested in array isn't supported, but could be if needed.

Like here it looks like you're doing just that - logging an array of records.  How are you expecting these to be consumed?  e.g. for our 1P partners who share their telemetry with us, in Azure DataExplorer - are you going to navigate through the dynamic object and do some aggregation across a whole matrix per summary?  Or what?

**Comment by markfields** [comment]
@jatgarg How are you planning to use this information?  And any concerns about scale?  I guess we're already logging the SummarizerTelemetry event every summary, it's just adding extra info.  `O(number of tables)`, right?.

File: packages/runtime/runtime-definitions/package.json [hunk 7]
```diff
@@ -63,7 +63,8 @@
 		"@fluidframework/core-interfaces": "workspace:~",
 		"@fluidframework/driver-definitions": "workspace:~",
 		"@fluidframework/id-compressor": "workspace:~",
-		"@fluidframework/protocol-definitions": "^3.1.0"
+		"@fluidframework/protocol-definitions": "^3.1.0",
+		"@fluidframework/telemetry-utils": "workspace:~"
```
Comment (markfields) [inline packages/runtime/runtime-definitions/package.json:67]: Did this pass layer-check??  Doesn't feel right to add this dependency from a `-definitions` package to a `-utils` package.

File: packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md [hunk 8]
```diff
@@ -397,11 +397,14 @@ export interface ISummaryTreeWithStats {
...
-    push(prefix: string, property: string, value: TelemetryEventPropertyType[]): void;
+    get(prefix: string, property: string): TelemetryEventPropertyTypeExt;
+    push(prefix: string, property: string, value: string | number | boolean | undefined | {
+        [key: string]: // Flat objects can have the same properties as the event itself
+        string | number | boolean | undefined | (string | number | boolean)[];
+    }): void;
     serialize(): string;
-    set(prefix: string, property: string, value: TelemetryEventPropertyType): void;
-    setMultiple(prefix: string, property: string, values: Record<string, TelemetryEventPropertyType>): void;
+    set(prefix: string, property: string, value: TelemetryEventPropertyTypeExt): void;
```
Comment (markfields) [inline packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md:406]: We want to avoid exposing the `Ext` ones on any public API if possible.

...Shoot I see some already are public.

File: packages/utils/telemetry-utils/api-report/telemetry-utils.api.md [hunk 9]
```diff
@@ -396,7 +396,10 @@ export enum TelemetryDataTag {
 export type TelemetryEventCategory = "generic" | "error" | "performance";
 
 // @public
-export type TelemetryEventPropertyTypeExt = string | number | boolean | undefined | (string | number | boolean)[] | {
+export type TelemetryEventPropertyTypeExt = string | number | boolean | undefined | (string | number | boolean | undefined | {
```
Comment (markfields) [inline packages/utils/telemetry-utils/api-report/telemetry-utils.api.md:399]: We should do this change in a separate PR to highlight and isolate it from the rest of this change.

Comment (markfields) [reply packages/utils/telemetry-utils/api-report/telemetry-utils.api.md:399]: And what I would say in _that_ PR is if we're doing this we might as well go full-on `Jsonable`.

File: packages/runtime/runtime-definitions/package.json (hunk 7 above)
Comment (jatgarg) [reply packages/runtime/runtime-definitions/package.json:67]: Yes, it passes the check. Since it is not the core defn, these already depends on the utils in layerinfo.json.

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 10]
```diff
@@ -346,7 +356,7 @@ export interface ITelemetryContext {
 	 * @param property - property name of the telemetry data being tracked (ex: "DirectoryCount")
 	 * @returns undefined if item not found
 	 */
-	get(prefix: string, property: string): TelemetryEventPropertyType;
+	get(prefix: string, property: string): TelemetryEventPropertyTypeExt;
```
Comment (markfields) [inline packages/runtime/runtime-definitions/src/summary.ts:359]: I wish we could avoid exposing the `Ext` types on any public API.  Ideally `ITelemetryContext` wouldn't be public but it currently has to be.

The reason that's a bummer is that then we can't easily expand the supported types (like we discussed allowing arbitrary Jsonable).  _If_ it's only used as an "in" type (users of public API pass _in_ the telemetry event) then we actually can expand what it accepts and it'll be automatically backwards compatible (can do in a minor bump).  So I'd like to keep `get` here as returning the non-`Ext` one for now.  I am gonna put out a small PR to deprecate it, actually. This should be a "write-only" API, like loggers (they only have `send`).

Comment (markfields) [reply packages/runtime/runtime-definitions/src/summary.ts:359]: PR to deprecate get/serialize on here:  #19409

For now, can you avoid changing the return type of `get`? You can do something quick to convert like `return typeof value === "object" ? JSON.stringify(value) : value;` (as opposed to the big switch statement and throwing in the default case etc - it just won't happen)

File: packages/runtime/runtime-utils/src/summaryUtils.ts [hunk 11]
```diff
@@ -379,22 +379,36 @@ export function convertSummaryTreeToITree(summaryTree: ISummaryTree): ITree {
...
 	 */
-	set(prefix: string, property: string, value: TelemetryEventPropertyType): void {
+	set(prefix: string, property: string, value: TelemetryEventPropertyTypeExt): void {
 		this.telemetry.set(`${prefix}${property}`, value);
 	}
 
+	/**
+	 * {@inheritDoc @fluidframework/runtime-definitions#ITelemetryContext.push}
+	 */
+	push(prefix: string, property: string, value: TelemetryEventPropertyTypeExt): void {
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/runtime-utils/src/summaryUtils.ts:394]: If you do keep it as-is, please put some big comments somewhere (including beside the cast) explaining what's going on and why it's ok

File: packages/dds/matrix/src/matrix.ts [hunk 12]
```diff
@@ -536,20 +539,31 @@ export class SharedMatrix<T = any>
...
+			artifactsToSummarize.push(trackerMatrixSnapshot);
+			props.tmsSize = trackerMatrixSnapshot.length; // trackerMatrixSnapshotSize
 		}
 		builder.addBlob(
 			SnapshotPath.cells,
 			serializer.stringify(artifactsToSummarize, this.handle),
 		);
+		if (telemetryContext?.push !== undefined) {
+			telemetryContext.push("fluid:SharedMatrix", "details", props);
+		}
```
Comment (markfields) [inline packages/dds/matrix/src/matrix.ts:566]: Are you checking `push` for `undefined` for some back-compat case? I guess old ContainerRuntime, new DataStoreRuntime?

I prefer we codify this in the interface itself, declaring `push` to be optional... but since it's a public interface, it may be stuck as optional until `3.0`? That's fine I guess.  Or if it goes into rc.2.0 as optional maybe in 2.0.0 it can be required. Don't spend a lot of time sorting it out, but do consider making it optional on the interface to get the compiler's help (no one else using it will think to check for `undefined` first otherwise!)

You can in-line the `undefined` check on the function like this:
```suggestion
		telemetryContext?.push?.("fluid:SharedMatrix", "details", props);
```

**Review by markfields — CHANGES_REQUESTED** [review]
Couple last tweaks around the types, but otherwise looks good!

File: packages/runtime/runtime-utils/package.json [hunk 13]
```diff
@@ -166,6 +166,10 @@
 		}
 	},
 	"typeValidation": {
-		"broken": {}
+		"broken": {
+			"ClassDeclaration_TelemetryContext": {
+				"forwardCompat": false
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/runtime/runtime-utils/package.json:171]: Removed it by making push optional.
